From 3a2e3b5d9571b86f8856004db14dc60853b09f0f Mon Sep 17 00:00:00 2001
From: Frogging-Family
From: Tk-Glitch <ti3nou@gmail.com>
From: Adel KARA SLIMANE <adel.ks@zegrapher.com>
From: damachine <damachin3@proton.me>
Date: Fri, 20 Mar 2026 19:46:07 +0100
Subject: [PATCH] linux-tkg: tk-glitch 0013-optimize_harder_O3.patch for 6.18

Signed-off-by: patch-family-tool <patches@frogging-family.invalid>
---
 Makefile     |  8 ++++++++
 init/Kconfig | 26 ++++++++++++++++++++++++++
 2 files changed, 34 insertions(+)

diff --git a/Makefile b/Makefile
index 67c2f5dbb..9ceb369a5 100644
--- a/Makefile
+++ b/Makefile
@@ -870,11 +870,19 @@ KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
 KBUILD_RUSTFLAGS += -Copt-level=2
+else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
+KBUILD_CFLAGS += -O3
+KBUILD_RUSTFLAGS += -Copt-level=3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
 KBUILD_RUSTFLAGS += -Copt-level=s
 endif
 
+# Perform swing modulo scheduling immediately before the first scheduling pass.
+# This pass looks at innermost loops and reorders their instructions by
+# overlapping different iterations.
+KBUILD_CFLAGS += $(call cc-option,-fmodulo-sched -fmodulo-sched-allow-regmoves -fivopts)
+
 # Always set `debug-assertions` and `overflow-checks` because their default
 # depends on `opt-level` and `debug-assertions`, respectively.
 KBUILD_RUSTFLAGS += -Cdebug-assertions=$(if $(CONFIG_RUST_DEBUG_ASSERTIONS),y,n)
diff --git a/init/Kconfig b/init/Kconfig
index cab3ad28c..3a664a42c 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1541,6 +1561,12 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 	  with the "-O2" compiler flag for best performance and most
 	  helpful compile-time warnings.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE_O3
+	bool "Optimize more for performance (-O3)"
+	help
+	  Choosing this option will pass "-O3" to your compiler to optimize
+	  the kernel yet more for performance.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
 	help
-- 
2.53.0

