From 6368e2994fb1d82dd159f6b5ef021adaf92912a4 Mon Sep 17 00:00:00 2001
From: Hyper-KVM <hyperkvmx86@gmail.com>
Date: Mon, 12 Oct 2020 21:05:18 -0400
Subject: [PATCH] scripts/mkdebian: remove build depends

---
 scripts/package/mkdebian | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/package/mkdebian b/scripts/package/mkdebian
index 48fbd3d0284a..c0f93dd33d74 100755
--- a/scripts/package/mkdebian
+++ b/scripts/package/mkdebian
@@ -174,7 +174,7 @@ Source: $sourcename
 Section: kernel
 Priority: optional
 Maintainer: $maintainer
-Build-Depends: bc, rsync, kmod, cpio, bison, flex | flex:native $extra_build_depends
+Build-Depends:
 Homepage: https://www.kernel.org/
 
 Package: $packagename
-- 
2.28.0

