diff --git a/scripts/package/mkspec b/scripts/package/mkspec
index f892cf8e37..d06fefbfdf 100755
--- a/scripts/package/mkspec
+++ b/scripts/package/mkspec
@@ -23,11 +23,11 @@ else
 	M=DEL
 fi
 
+PROVIDES_DRM=""
 if grep -q CONFIG_DRM=y .config; then
-	PROVIDES=kernel-drm
+	PROVIDES_DRM="Provides: kernel-drm = %{version}"
 fi
 
-PROVIDES="$PROVIDES kernel-$KERNELRELEASE"
 __KERNELRELEASE=$(echo $KERNELRELEASE | sed -e "s/-/_/g")
 EXCLUDES="$RCS_TAR_IGNORE --exclude=*vmlinux* --exclude=*.mod \
 --exclude=*.o --exclude=*.ko --exclude=*.cmd --exclude=Documentation \
@@ -48,7 +48,10 @@ sed -e '/^DEL/d' -e 's/^\t*//' <<EOF
 	Vendor: The Linux Community
 	URL: https://www.kernel.org
 $S	Source: kernel-$__KERNELRELEASE.tar.gz
-	Provides: $PROVIDES
+	$PROVIDES_DRM
+	Provides: kernel = %{version}
+	Provides: kernel-uname-r = %{version}
+	Provides: installonlypkg(kernel) = %{version}
 	# $UTS_MACHINE as a fallback of _arch in case
 	# /usr/lib/rpm/platform/*/macros was not included.
 	%{!?_arch: %define _arch $UTS_MACHINE}
@@ -61,8 +64,8 @@ $S	Source: kernel-$__KERNELRELEASE.tar.gz
 	%package headers
 	Summary: Header files for the Linux kernel for use by glibc
 	Group: Development/System
-	Obsoletes: kernel-headers
 	Provides: kernel-headers = %{version}
+	Provides: installonlypkg(kernel) = %{version}
 	%description headers
 	Kernel-headers includes the C header files that specify the interface
 	between the Linux kernel and userspace libraries and programs.  The
@@ -73,11 +76,30 @@ $S	Source: kernel-$__KERNELRELEASE.tar.gz
 $S$M	%package devel
 $S$M	Summary: Development package for building kernel modules to match the $__KERNELRELEASE kernel
 $S$M	Group: System Environment/Kernel
+$S$M	Provides: kernel-devel = %{version}
+$S$M	Provides: kernel-devel-uname-r = %{version}
+$S$M	Provides: installonlypkg(kernel) = %{version}
 $S$M	AutoReqProv: no
 $S$M	%description -n kernel-devel
 $S$M	This package provides kernel headers and makefiles sufficient to build modules
 $S$M	against the $__KERNELRELEASE kernel package.
 $S$M
+$S	# Opt out of a lot of Fedora hardening flags etc...
+$S	# See https://src.fedoraproject.org/rpms/redhat-rpm-config//blob/rawhide/f/buildflags.md
+$S	%undefine _package_note_file
+$S	%undefine _auto_set_build_flags
+$S	%undefine _include_frame_pointers
+$S	%define _build_id_flags -Wl,--build-id=none
+$S	%undefine _annotated_build
+$S	%undefine _fortify_level
+$S	%undefine _hardened_build
+$S	%global _lto_cflags %{nil}
+$S	%global _configure_gnuconfig_hack 0
+$S	%global _configure_libtool_hardening_hack 0
+$S	# Nearly had to go to the deep web to find documentation on this one... Gosh
+$S	# See https://github.com/rpm-software-management/rpm/blob/master/macros.in#L471
+$S	%define _build_id_links none
+$S
 $S	%prep
 $S	%setup -q
 $S
