From f7f49141a5dbe9c99d78196b58c44307fb2e6be3 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Wed, 4 Jul 2018 04:30:08 +0200
Subject: glitched - MuQSS

diff --git a/kernel/sched/MuQSS.c b/kernel/sched/MuQSS.c
index 84a1d08d68551..57c3036a68952 100644
--- a/kernel/sched/MuQSS.c
+++ b/kernel/sched/MuQSS.c
@@ -163,7 +167,11 @@ int sched_interactive __read_mostly = 1;
  * are allowed to run five seconds as real time tasks. This is the total over
  * all online cpus.
  */
+#ifdef CONFIG_ZENIFY
+int sched_iso_cpu __read_mostly = 25;
+#else
 int sched_iso_cpu __read_mostly = 70;
+#endif
 
 /*
  * sched_yield_type - Choose what sort of yield sched_yield will perform.

diff --git a/kernel/Kconfig.hz b/kernel/Kconfig.hz
index 2a202a846757..1d9c7ed79b11 100644
--- a/kernel/Kconfig.hz
+++ b/kernel/Kconfig.hz
@@ -5,7 +5,7 @@
 choice
 	prompt "Timer frequency"
 	default HZ_100 if SCHED_MUQSS
-	default HZ_250_NODEF if !SCHED_MUQSS
+	default HZ_500_NODEF if !SCHED_MUQSS
 	help
 	 Allows the configuration of the timer frequency. It is customary
 	 to have the timer interrupt run at 1000 Hz but 100 Hz may be more
@@ -50,6 +50,20 @@ choice
 	 on SMP and NUMA systems and exactly dividing by both PAL and
 	 NTSC frame rates for video and multimedia work.
 
+	config HZ_500_NODEF
+		bool "500 HZ"
+	help
+	 500 Hz is a good timer frequency for desktops. Provides fast
+	 interactivity with great smoothness without sacrificing too
+	 much throughput.
+
+	config HZ_750_NODEF
+		bool "750 HZ"
+	help
+	 750 Hz is a good timer frequency for desktops. Provides fast
+	 interactivity with great smoothness without sacrificing too
+	 much throughput.
+
 	config HZ_1000_NODEF
 		bool "1000 HZ"
 	help
@@ -63,6 +70,8 @@ config HZ
 	default 100 if HZ_100
 	default 250 if HZ_250_NODEF
 	default 300 if HZ_300_NODEF
+	default 500 if HZ_500_NODEF
+	default 750 if HZ_750_NODEF
 	default 1000 if HZ_1000_NODEF
 
 config SCHED_HRTICK

diff --git a/Makefile b/Makefile
index d4d36c61940b..4a9dfe471f1f 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,6 @@ NAME = Kleptomaniac Octopus
 
 CKVERSION = -ck1
 CKNAME = MuQSS Powered
-EXTRAVERSION := $(EXTRAVERSION)$(CKVERSION)
 
 # We are using a recursive build, so we need to do a little thinking
 # to get the ordering right.
diff --git a/scripts/headers_install.sh b/scripts/headers_install.sh
index dd554bd43..75030ad93 100755
--- a/scripts/headers_install.sh
+++ b/scripts/headers_install.sh
@@ -89,6 +89,7 @@ include/uapi/linux/atmdev.h:CONFIG_COMPAT
 include/uapi/linux/eventpoll.h:CONFIG_PM_SLEEP
 include/uapi/linux/hw_breakpoint.h:CONFIG_HAVE_MIXED_BREAKPOINTS_REGS
 include/uapi/linux/pktcdvd.h:CONFIG_CDROM_PKTCDVD_WCACHE
+include/uapi/linux/sched.h:CONFIG_SCHED_MUQSS
 "
 
 for c in $configs
