--- a/drivers/base/firmware_loader/main.c	2021-08-24 15:42:07.025482085 -0400
+++ b/drivers/base/firmware_loader/main.c	2021-08-24 15:44:40.782975313 -0400
@@ -809,6 +809,11 @@ _request_firmware(const struct firmware
 
 	ret = _request_firmware_prepare(&fw, name, device, buf, size,
 					offset, opt_flags);
+
+#ifdef CONFIG_GENTOO_PRINT_FIRMWARE_INFO
+        printk(KERN_NOTICE "Loading firmware: %s\n", name);
+#endif
+
 	if (ret <= 0) /* error or already assigned */
 		goto out;
 
