diff --git a/scripts/package/mkspec b/scripts/package/mkspec
index 8fa7c5b8a..cf6fc13f1 100755
--- a/scripts/package/mkspec
+++ b/scripts/package/mkspec
@@ -24,10 +24,9 @@ else
 fi
 
 if grep -q CONFIG_DRM=y .config; then
-	PROVIDES=kernel-drm
+	PROVIDES_DRM="Provides: kernel-drm = %{version}"
 fi
 
-PROVIDES="$PROVIDES kernel-$KERNELRELEASE"
 __KERNELRELEASE=$(echo $KERNELRELEASE | sed -e "s/-/_/g")
 EXCLUDES="$RCS_TAR_IGNORE --exclude=*vmlinux* --exclude=*.mod \
 --exclude=*.o --exclude=*.ko --exclude=*.cmd --exclude=Documentation \
@@ -48,7 +47,9 @@ sed -e '/^DEL/d' -e 's/^\t*//' <<EOF
 	Vendor: The Linux Community
 	URL: https://www.kernel.org
 $S	Source: kernel-$__KERNELRELEASE.tar.gz
-	Provides: $PROVIDES
+	$PROVIDES_DRM
+	Provides: kernel = %{version}
+	Provides: installonlypkg(kernel) = %{version}
 	# $UTS_MACHINE as a fallback of _arch in case
 	# /usr/lib/rpm/platform/*/macros was not included.
 	%define _arch %{?_arch:$UTS_MACHINE}
@@ -61,7 +62,7 @@ $S	Source: kernel-$__KERNELRELEASE.tar.gz
 	%package headers
 	Summary: Header files for the Linux kernel for use by glibc
 	Group: Development/System
-	Obsoletes: kernel-headers
+	Provides: installonlypkg(kernel) = %{version}
 	Provides: kernel-headers = %{version}
 	%description headers
 	Kernel-headers includes the C header files that specify the interface
@@ -73,6 +74,9 @@ $S	Source: kernel-$__KERNELRELEASE.tar.gz
 $S$M	%package devel
 $S$M	Summary: Development package for building kernel modules to match the $__KERNELRELEASE kernel
 $S$M	Group: System Environment/Kernel
+$S$M	Provides: kernel-devel = %{version}
+$S$M	Provides: kernel-devel-uname-r = %{version}
+$S$M	Provides: installonlypkg(kernel) = %{version}
 $S$M	AutoReqProv: no
 $S$M	%description -n kernel-devel
 $S$M	This package provides kernel headers and makefiles sufficient to build modules
