diff --git a/scripts/package/kernel.spec b/scripts/package/kernel.spec
index 3eee0143e0..bd1a95c897 100644
--- a/scripts/package/kernel.spec
+++ b/scripts/package/kernel.spec
@@ -16,6 +16,7 @@ URL: https://www.kernel.org
 Source0: linux.tar.gz
 Source1: config
 Source2: diff.patch
+Provides: kernel-uname-r = %{version}
 Provides: kernel-%{KERNELRELEASE}
 BuildRequires: bc binutils bison dwarves
 BuildRequires: (elfutils-libelf-devel or libelf-devel) flex
@@ -27,8 +28,8 @@ The Linux Kernel, the operating system core itself
 %package headers
 Summary: Header files for the Linux kernel for use by glibc
 Group: Development/System
-Obsoletes: kernel-headers
 Provides: kernel-headers = %{version}
+Provides: installonlypkg(kernel) = %{version}
 %description headers
 Kernel-headers includes the C header files that specify the interface
 between the Linux kernel and userspace libraries and programs.  The
@@ -40,12 +41,31 @@ glibc package.
 %package devel
 Summary: Development package for building kernel modules to match the %{version} kernel
 Group: System Environment/Kernel
+Provides: kernel-devel = %{version}
+Provides: kernel-devel-uname-r = %{version}
+Provides: installonlypkg(kernel) = %{version}
 AutoReqProv: no
 %description -n kernel-devel
 This package provides kernel headers and makefiles sufficient to build modules
 against the %{version} kernel package.
 %endif
 
+# Opt out of a lot of Fedora hardening flags etc...
+# See https://src.fedoraproject.org/rpms/redhat-rpm-config//blob/rawhide/f/buildflags.md
+%undefine _package_note_file
+%undefine _auto_set_build_flags
+%undefine _include_frame_pointers
+%define _build_id_flags -Wl,--build-id=none
+%undefine _annotated_build
+%undefine _fortify_level
+%undefine _hardened_build
+%global _lto_cflags %{nil}
+%global _configure_gnuconfig_hack 0
+%global _configure_libtool_hardening_hack 0
+# Nearly had to go to the deep web to find documentation on this one... Gosh
+# See https://github.com/rpm-software-management/rpm/blob/master/macros.in#L471
+%define _build_id_links none
+
 %prep
 %setup -q -n linux
 cp %{SOURCE1} .config
