From 24cd48424c96118bbb0fc8af93bfdc98570c31bf Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 25 Oct 2023 21:49:14 +0200
Subject: [PATCH] EEVDF: Disable DELAY_DEQUEUE

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/features.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/features.h b/kernel/sched/features.h
index fd2c963b7..6225be566 100644
--- a/kernel/sched/features.h
+++ b/kernel/sched/features.h
@@ -44,7 +44,7 @@ SCHED_FEAT(EVDF, false)
  *
  * GENTLE_DELAY clips the lag on dequeue (or wakeup) to 0.
  */
-SCHED_FEAT(DELAY_DEQUEUE, true)
+SCHED_FEAT(DELAY_DEQUEUE, false)
 SCHED_FEAT(GENTLE_DELAY, true)
 
 /*
-- 
2.42.0

